File: tools/export_asset.py
#!/usr/bin/env python3
"""Export exact, palette-independent GBA tile indices to indexed PNG."""
import argparse
import json
import struct
import zlib
from pathlib import Path

from import_asset import gba_graphics, gba_palette_rgba, indexed_png, rgba_png


ROM_BASE = 0x08000000


def chunk(kind, payload):
    body = kind + payload
    return (struct.pack(">I", len(payload)) + body +
            struct.pack(">I", zlib.crc32(body) & 0xffffffff))


def tile_png(raw, bpp, columns, palette_colors=None):
    unit = 32 if bpp == 4 else 64
    if not raw or len(raw) % unit:
        raise ValueError("tile data must contain whole nonempty tiles")
    count = len(raw) // unit
    if columns <= 0 or count % columns:
        raise ValueError("columns must divide the tile count exactly")
    rows = count // columns
    width, height = columns * 8, rows * 8
    pixels = bytearray(width * height)
    for tile in range(count):
        source = raw[tile * unit:(tile + 1) * unit]
        indices = (
            [value for byte in source for value in (byte & 15, byte >> 4)]
            if bpp == 4 else list(source)
        )
        left = tile % columns * 8
        top = tile // columns * 8
        for y in range(8):
            start = (top + y) * width + left
            pixels[start:start + 8] = bytes(indices[y * 8:y * 8 + 8])
    scanlines = bytearray()
    for y in range(height):
        row = pixels[y * width:(y + 1) * width]
        scanlines.append(0)
        if bpp == 4:
            scanlines.extend(row[index] << 4 | row[index + 1]
                             for index in range(0, width, 2))
        else:
            scanlines.extend(row)
    colors = 16 if bpp == 4 else 256
    if palette_colors is None:
        palette_colors = (
            [(index * 16,) * 3 for index in range(16)]
            if bpp == 4 else
            [((index & 31) * 8, (index >> 5) * 8, 0)
             for index in range(256)]
        )
    if not 1 <= len(palette_colors) <= colors:
        raise ValueError("palette does not fit the requested tile depth")
    if max(pixels) >= len(palette_colors):
        raise ValueError("tile pixels reference a missing palette entry")
    palette = bytes(channel for color in palette_colors for channel in color)
    png = (
        b"\x89PNG\r\n\x1a\n" +
        chunk(b"IHDR", struct.pack(">IIBBBBB",
                                    width, height, bpp, 3, 0, 0, 0)) +
        chunk(b"PLTE", palette) +
        chunk(b"IDAT", zlib.compress(bytes(scanlines), 9)) +
        chunk(b"IEND", b"")
    )
    return png, {"width": width, "height": height, "bpp": bpp,
                 "tiles": count, "columns": columns}


def palette_png(raw):
    if not raw or len(raw) % 2 or len(raw) > 512:
        raise ValueError("palette must contain 1..256 BGR555 entries")
    entries = len(raw) // 2
    palette = bytearray()
    for offset in range(0, len(raw), 2):
        value = struct.unpack_from("<H", raw, offset)[0]
        if value & 0x8000:
            raise ValueError("palette contains a non-BGR555 high bit")
        palette.extend(((value & 31) << 3,
                        (value >> 5 & 31) << 3,
                        (value >> 10 & 31) << 3))
    pixels = list(range(entries)) + [0] * (256 - entries)
    scanlines = b"".join(
        b"\0" + bytes(pixels[row * 16:(row + 1) * 16])
        for row in range(16))
    png = (
        b"\x89PNG\r\n\x1a\n" +
        chunk(b"IHDR", struct.pack(">IIBBBBB", 16, 16, 8, 3, 0, 0, 0)) +
        chunk(b"PLTE", bytes(palette)) +
        chunk(b"IDAT", zlib.compress(scanlines, 9)) +
        chunk(b"IEND", b"")
    )
    return png, {"width": 16, "height": 16,
                 "palette_entries": entries}


def byte_png(raw, width):
    if (not raw or width <= 0 or len(raw) % width or width % 8 or
            (len(raw) // width) % 8):
        raise ValueError("byte image dimensions must be nonzero multiples of eight")
    height = len(raw) // width
    palette = bytes(channel for index in range(256)
                    for channel in (index, index, index))
    scanlines = b"".join(
        b"\0" + raw[row * width:(row + 1) * width]
        for row in range(height))
    png = (
        b"\x89PNG\r\n\x1a\n" +
        chunk(b"IHDR", struct.pack(">IIBBBBB",
                                    width, height, 8, 3, 0, 0, 0)) +
        chunk(b"PLTE", palette) +
        chunk(b"IDAT", zlib.compress(scanlines, 9)) +
        chunk(b"IEND", b"")
    )
    return png, {"width": width, "height": height, "bytes": len(raw)}


def rgba_image(raw, width):
    if not raw or len(raw) % 4 or width <= 0:
        raise ValueError("RGBA input must contain whole nonempty pixels")
    pixels = len(raw) // 4
    if pixels % width:
        raise ValueError("width must divide the RGBA pixel count")
    height = pixels // width
    scanlines = b"".join(
        b"\0" + raw[row * width * 4:(row + 1) * width * 4]
        for row in range(height))
    png = (
        b"\x89PNG\r\n\x1a\n" +
        chunk(b"IHDR", struct.pack(">IIBBBBB",
                                    width, height, 8, 6, 0, 0, 0)) +
        chunk(b"IDAT", zlib.compress(scanlines, 9)) +
        chunk(b"IEND", b"")
    )
    return png, {"width": width, "height": height, "pixels": pixels}


def palette_rgba_image(raw, width):
    if not raw or len(raw) % 2 or width <= 0 or len(raw) // 2 % width:
        raise ValueError("palette RGBA dimensions must divide whole colors")
    pixels = bytearray()
    for offset in range(0, len(raw), 2):
        value = struct.unpack_from("<H", raw, offset)[0]
        pixels.extend(((value & 31) << 3, (value >> 5 & 31) << 3,
                       (value >> 10 & 31) << 3,
                       254 if value & 0x8000 else 255))
    png, report = rgba_image(bytes(pixels), width)
    report["palette_entries"] = len(raw) // 2
    return png, report


def self_test():
    for bpp, size in ((4, 32 * 7), (8, 64 * 4)):
        raw = bytes((index * 37 + 11) & 255 for index in range(size))
        png, _ = tile_png(raw, bpp, 7 if bpp == 4 else 4)
        tiles, _, _ = gba_graphics(png, bpp)
        if tiles != raw:
            raise AssertionError(f"{bpp}bpp tile round-trip failed")
    colors = [((index & 31) << 3, (index >> 3 & 31) << 3,
               (index >> 6 & 3) << 3) for index in range(256)]
    raw = bytes((index * 53 + 7) & 255 for index in range(64 * 4))
    png, _ = tile_png(raw, 8, 4, colors)
    tiles, palette, _ = gba_graphics(png, 8)
    expected_palette = b"".join(
        struct.pack("<H", (red >> 3) | (green >> 3) << 5 |
                    (blue >> 3) << 10) for red, green, blue in colors)
    if tiles != raw or palette != expected_palette:
        raise AssertionError("palette-correct tile round-trip failed")
    raw = b"".join(struct.pack("<H", (index * 109) & 0x7fff)
                   for index in range(224))
    png, _ = palette_png(raw)
    _, palette, _ = gba_graphics(png, 8)
    if palette != raw:
        raise AssertionError("palette round-trip failed")
    raw = bytes((index * 53 + 7) & 255 for index in range(128 * 64))
    png, _ = byte_png(raw, 128)
    _, _, pixels, _ = indexed_png(png)
    if bytes(pixels) != raw:
        raise AssertionError("indexed byte-image round-trip failed")
    raw = bytes((index * 29 + 3) & 255 for index in range(11 * 7 * 4))
    png, _ = rgba_image(raw, 11)
    if rgba_png(png) != (11, 7, raw):
        raise AssertionError("RGBA image round-trip failed")
    raw = b"".join(struct.pack("<H", (index * 109) & 0xffff)
                   for index in range(128))
    png, _ = palette_rgba_image(raw, 16)
    if gba_palette_rgba(png)[0] != raw:
        raise AssertionError("RGBA palette round-trip failed")
    print("self-test=ok")


def number(text):
    try:
        return int(text, 0)
    except ValueError as error:
        raise argparse.ArgumentTypeError(f"invalid integer: {text}") from error


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--self-test", action="store_true")
    commands = parser.add_subparsers(dest="command")
    tiles = commands.add_parser("tiles")
    tiles.add_argument("rom", type=Path)
    tiles.add_argument("--address", type=number, required=True)
    tiles.add_argument("--size", type=number, required=True)
    tiles.add_argument("--bpp", type=int, choices=(4, 8), required=True)
    tiles.add_argument("--columns", type=int, required=True)
    tiles.add_argument("--palette", type=Path)
    tiles.add_argument("-o", "--output", type=Path, required=True)
    tile_file = commands.add_parser("tiles-file")
    tile_file.add_argument("input", type=Path)
    tile_file.add_argument("--bpp", type=int, choices=(4, 8), required=True)
    tile_file.add_argument("--columns", type=int, required=True)
    tile_file.add_argument("--palette", type=Path)
    tile_file.add_argument("-o", "--output", type=Path, required=True)
    palette = commands.add_parser("palette-file")
    palette.add_argument("input", type=Path)
    palette.add_argument("-o", "--output", type=Path, required=True)
    byte_file = commands.add_parser("bytes-file")
    byte_file.add_argument("input", type=Path)
    byte_file.add_argument("--width", type=int, required=True)
    byte_file.add_argument("-o", "--output", type=Path, required=True)
    rgba_file = commands.add_parser("rgba-file")
    rgba_file.add_argument("input", type=Path)
    rgba_file.add_argument("--width", type=int, required=True)
    rgba_file.add_argument("-o", "--output", type=Path, required=True)
    rgba_palette = commands.add_parser("palette-rgba-file")
    rgba_palette.add_argument("input", type=Path)
    rgba_palette.add_argument("--width", type=int, default=16)
    rgba_palette.add_argument("-o", "--output", type=Path, required=True)
    rgba_palette_rom = commands.add_parser("palette-rgba")
    rgba_palette_rom.add_argument("rom", type=Path)
    rgba_palette_rom.add_argument("--address", type=number, required=True)
    rgba_palette_rom.add_argument("--size", type=number, required=True)
    rgba_palette_rom.add_argument("--width", type=int, default=16)
    rgba_palette_rom.add_argument("-o", "--output", type=Path, required=True)
    args = parser.parse_args()
    if args.self_test:
        self_test()
        if args.command is None:
            return
    if args.command == "tiles":
        rom = args.rom.read_bytes()
        start = args.address - ROM_BASE
        end = start + args.size
        if start < 0 or end > len(rom) or start >= end:
            parser.error("tile range is outside the ROM or empty")
        if args.output.resolve() == args.rom.resolve():
            parser.error("refusing to overwrite the input ROM")
        colors = (indexed_png(args.palette.read_bytes())[3]
                  if args.palette is not None else None)
        png, report = tile_png(
            rom[start:end], args.bpp, args.columns, colors)
    elif args.command == "tiles-file":
        if args.output.resolve() == args.input.resolve():
            parser.error("refusing to overwrite the input")
        colors = (indexed_png(args.palette.read_bytes())[3]
                  if args.palette is not None else None)
        png, report = tile_png(
            args.input.read_bytes(), args.bpp, args.columns, colors)
    elif args.command == "palette-file":
        if args.output.resolve() == args.input.resolve():
            parser.error("refusing to overwrite the input")
        png, report = palette_png(args.input.read_bytes())
    elif args.command == "bytes-file":
        if args.output.resolve() == args.input.resolve():
            parser.error("refusing to overwrite the input")
        png, report = byte_png(args.input.read_bytes(), args.width)
    elif args.command == "rgba-file":
        if args.output.resolve() == args.input.resolve():
            parser.error("refusing to overwrite the input")
        png, report = rgba_image(args.input.read_bytes(), args.width)
    elif args.command == "palette-rgba-file":
        if args.output.resolve() == args.input.resolve():
            parser.error("refusing to overwrite the input")
        png, report = palette_rgba_image(
            args.input.read_bytes(), args.width)
    elif args.command == "palette-rgba":
        rom = args.rom.read_bytes()
        start = args.address - ROM_BASE
        end = start + args.size
        if start < 0 or end > len(rom) or start >= end:
            parser.error("palette range is outside the ROM or empty")
        if args.output.resolve() == args.rom.resolve():
            parser.error("refusing to overwrite the input ROM")
        png, report = palette_rgba_image(rom[start:end], args.width)
    else:
        parser.error("an asset command is required")
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_bytes(png)
    print(json.dumps(report, sort_keys=True))


if __name__ == "__main__":
    main()

File: tools/export_asset.ts
// Tool role: both; imported by tools/archive_asset.ts, tools/battle_effect_data.ts, tools/chiiki_map_resources.ts (+18 more); invoked by package.json.
import { mkdirSync, readFileSync, realpathSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { crc32, deflateSync } from "./zlib.ts";
import { gba_graphics, gba_palette_rgba, indexed_png, rgba_png, type Rgb, type Report } from "./import_asset.ts";

export const ROM_BASE = 0x08000000;

export function chunk(kind: Uint8Array, payload: Uint8Array): Buffer {
  const body = Buffer.concat([Buffer.from(kind), Buffer.from(payload)]);
  const size = Buffer.alloc(4), checksum = Buffer.alloc(4);
  size.writeUInt32BE(payload.length); checksum.writeUInt32BE(crc32(body));
  return Buffer.concat([size, body, checksum]);
}

function png(parts: Buffer[]): Buffer {
  return Buffer.concat([Buffer.from("89504e470d0a1a0a", "hex"), ...parts]);
}

function ihdr(width: number, height: number, depth: number, color: number): Buffer {
  const value = Buffer.alloc(13);
  value.writeUInt32BE(width, 0); value.writeUInt32BE(height, 4);
  value.set([depth, color, 0, 0, 0], 8);
  return value;
}

export function tile_png(raw: Uint8Array, bpp: number, columns: number, palette_colors?: readonly Rgb[] | null): [Buffer, Report] {
  const unit = bpp === 4 ? 32 : 64;
  if (raw.length === 0 || raw.length % unit !== 0) throw new Error("tile data must contain whole nonempty tiles");
  const count = raw.length / unit;
  if (columns <= 0 || count % columns !== 0) throw new Error("columns must divide the tile count exactly");
  const rows = count / columns, width = columns * 8, height = rows * 8;
  const pixels = Buffer.alloc(width * height);
  for (let tile = 0; tile < count; tile++) {
    const source = raw.subarray(tile * unit, (tile + 1) * unit);
    const indices = bpp === 4 ? Array.from(source).flatMap((byte) => [byte & 15, byte >>> 4]) : Array.from(source);
    const left = tile % columns * 8, top = Math.floor(tile / columns) * 8;
    for (let y = 0; y < 8; y++) pixels.set(indices.slice(y * 8, y * 8 + 8), (top + y) * width + left);
  }
  const scanlines: number[] = [];
  for (let y = 0; y < height; y++) {
    scanlines.push(0);
    const row = pixels.subarray(y * width, (y + 1) * width);
    if (bpp === 4) for (let index = 0; index < width; index += 2) scanlines.push(row[index] << 4 | row[index + 1]);
    else scanlines.push(...row);
  }
  const colors = bpp === 4 ? 16 : 256;
  let palette: readonly Rgb[];
  if (palette_colors == null) {
    palette = bpp === 4 ? Array.from({ length: 16 }, (_, index) => [index * 16, index * 16, index * 16] as Rgb) :
      Array.from({ length: 256 }, (_, index) => [(index & 31) * 8, (index >>> 5) * 8, 0] as Rgb);
  } else palette = palette_colors;
  if (palette.length < 1 || palette.length > colors) throw new Error("palette does not fit the requested tile depth");
  if (Math.max(...pixels) >= palette.length) throw new Error("tile pixels reference a missing palette entry");
  const paletteBytes = Buffer.from(palette.flatMap((color) => color));
  return [png([chunk(Buffer.from("IHDR"), ihdr(width, height, bpp, 3)), chunk(Buffer.from("PLTE"), paletteBytes),
    chunk(Buffer.from("IDAT"), deflateSync(Buffer.from(scanlines), { level: 9 })), chunk(Buffer.from("IEND"), Buffer.alloc(0))]),
  { width, height, bpp, tiles: count, columns }];
}

export function palette_png(raw: Uint8Array): [Buffer, Report] {
  if (raw.length === 0 || raw.length % 2 !== 0 || raw.length > 512)
    throw new Error("palette must contain 1..256 BGR555 entries");
  const entries = raw.length / 2;
  const palette: number[] = [];
  const view = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  for (let offset = 0; offset < raw.length; offset += 2) {
    const value = view.getUint16(offset, true);
    if (value & 0x8000) throw new Error("palette contains a non-BGR555 high bit");
    palette.push((value & 31) << 3, (value >>> 5 & 31) << 3, (value >>> 10 & 31) << 3);
  }
  const pixels = [...Array.from({ length: entries }, (_, index) => index), ...Array(256 - entries).fill(0)];
  const scanlines: number[] = [];
  for (let row = 0; row < 16; row++) scanlines.push(0, ...pixels.slice(row * 16, (row + 1) * 16));
  return [png([chunk(Buffer.from("IHDR"), ihdr(16, 16, 8, 3)), chunk(Buffer.from("PLTE"), Buffer.from(palette)),
    chunk(Buffer.from("IDAT"), deflateSync(Buffer.from(scanlines), { level: 9 })), chunk(Buffer.from("IEND"), Buffer.alloc(0))]),
  { width: 16, height: 16, palette_entries: entries }];
}

export function byte_png(raw: Uint8Array, width: number): [Buffer, Report] {
  if (raw.length === 0 || width <= 0 || raw.length % width !== 0 || width % 8 !== 0 || (raw.length / width) % 8 !== 0)
    throw new Error("byte image dimensions must be nonzero multiples of eight");
  const height = raw.length / width;
  const palette = Buffer.from(Array.from({ length: 256 }, (_, index) => [index, index, index]).flat());
  const rows: Buffer[] = [];
  for (let row = 0; row < height; row++) rows.push(Buffer.from([0]), Buffer.from(raw.subarray(row * width, (row + 1) * width)));
  return [png([chunk(Buffer.from("IHDR"), ihdr(width, height, 8, 3)), chunk(Buffer.from("PLTE"), palette),
    chunk(Buffer.from("IDAT"), deflateSync(Buffer.concat(rows), { level: 9 })), chunk(Buffer.from("IEND"), Buffer.alloc(0))]),
  { width, height, bytes: raw.length }];
}

export function rgba_image(raw: Uint8Array, width: number): [Buffer, Report] {
  if (raw.length === 0 || raw.length % 4 !== 0 || width <= 0)
    throw new Error("RGBA input must contain whole nonempty pixels");
  const pixels = raw.length / 4;
  if (pixels % width !== 0) throw new Error("width must divide the RGBA pixel count");
  const height = pixels / width;
  const rows: Buffer[] = [];
  for (let row = 0; row < height; row++) rows.push(Buffer.from([0]), Buffer.from(raw.subarray(row * width * 4, (row + 1) * width * 4)));
  return [png([chunk(Buffer.from("IHDR"), ihdr(width, height, 8, 6)),
    chunk(Buffer.from("IDAT"), deflateSync(Buffer.concat(rows), { level: 9 })), chunk(Buffer.from("IEND"), Buffer.alloc(0))]),
  { width, height, pixels }];
}

export function palette_rgba_image(raw: Uint8Array, width: number): [Buffer, Report] {
  if (raw.length === 0 || raw.length % 2 !== 0 || width <= 0 || raw.length / 2 % width !== 0)
    throw new Error("palette RGBA dimensions must divide whole colors");
  const pixels = Buffer.alloc(raw.length * 2);
  const source = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  for (let offset = 0; offset < raw.length; offset += 2) {
    const value = source.getUint16(offset, true), target = offset * 2;
    pixels.set([(value & 31) << 3, (value >>> 5 & 31) << 3, (value >>> 10 & 31) << 3, value & 0x8000 ? 254 : 255], target);
  }
  const [image, report] = rgba_image(pixels, width);
  report.palette_entries = raw.length / 2;
  return [image, report];
}

export function self_test(): void {
  for (const [bpp, size, columns] of [[4, 32 * 7, 7], [8, 64 * 4, 4]]) {
    const raw = Buffer.from(Array.from({ length: size }, (_, index) => (index * 37 + 11) & 255));
    const [image] = tile_png(raw, bpp, columns);
    if (!gba_graphics(image, bpp)[0].equals(raw)) throw new Error(`${bpp}bpp tile round-trip failed`);
  }
  const colors = Array.from({ length: 256 }, (_, index) => [(index & 31) << 3, (index >>> 3 & 31) << 3, (index >>> 6 & 3) << 3] as Rgb);
  let raw = Buffer.from(Array.from({ length: 64 * 4 }, (_, index) => (index * 53 + 7) & 255));
  let [image] = tile_png(raw, 8, 4, colors);
  const [tiles, palette] = gba_graphics(image, 8);
  const expectedPalette = Buffer.alloc(512);
  colors.forEach(([red, green, blue], index) => expectedPalette.writeUInt16LE((red >>> 3) | (green >>> 3) << 5 | (blue >>> 3) << 10, index * 2));
  if (!tiles.equals(raw) || !palette.equals(expectedPalette)) throw new Error("palette-correct tile round-trip failed");
  raw = Buffer.alloc(448); for (let index = 0; index < 224; index++) raw.writeUInt16LE(index * 109 & 0x7fff, index * 2);
  [image] = palette_png(raw); if (!gba_graphics(image, 8)[1].equals(raw)) throw new Error("palette round-trip failed");
  raw = Buffer.from(Array.from({ length: 128 * 64 }, (_, index) => (index * 53 + 7) & 255));
  [image] = byte_png(raw, 128); if (!Buffer.from(indexed_png(image)[2]).equals(raw)) throw new Error("indexed byte-image round-trip failed");
  raw = Buffer.from(Array.from({ length: 11 * 7 * 4 }, (_, index) => (index * 29 + 3) & 255));
  [image] = rgba_image(raw, 11); const rgba = rgba_png(image);
  if (rgba[0] !== 11 || rgba[1] !== 7 || !rgba[2].equals(raw)) throw new Error("RGBA image round-trip failed");
  raw = Buffer.alloc(256); for (let index = 0; index < 128; index++) raw.writeUInt16LE(index * 109 & 0xffff, index * 2);
  [image] = palette_rgba_image(raw, 16); if (!gba_palette_rgba(image)[0].equals(raw)) throw new Error("RGBA palette round-trip failed");
  console.log("self-test=ok");
}

function option(args: string[], ...names: string[]): string {
  const index = args.findIndex((arg) => names.includes(arg));
  if (index < 0 || index + 1 >= args.length) throw new Error(`${names.at(-1)} is required`);
  return args[index + 1];
}

function number(text: string): number {
  const value = Number(text);
  if (!Number.isInteger(value)) throw new Error(`invalid integer: ${text}`);
  return value;
}

function same(a: string, b: string): boolean {
  try { return realpathSync(a) === realpathSync(b); } catch { return resolve(a) === resolve(b); }
}

function sortedJson(value: Record<string, unknown>): string {
  const sorted = Object.fromEntries(Object.keys(value).sort().map((key) => [key, value[key]]));
  return JSON.stringify(sorted).replace(/:/g, ": ").replace(/,/g, ", ");
}

function main(args: string[]): void {
  if (args.includes("-h") || args.includes("--help")) {
    console.log("usage: export_asset.ts [--self-test] {tiles,tiles-file,palette-file,bytes-file,rgba-file,palette-rgba-file,palette-rgba} ...");
    return;
  }
  if (args.includes("--self-test")) {
    self_test(); args = args.filter((arg) => arg !== "--self-test"); if (args.length === 0) return;
  }
  const command = args[0], input = args[1];
  if (!command || !input) throw new Error("an asset command is required");
  const output = option(args, "-o", "--output");
  let image: Buffer, report: Report;
  if (command === "tiles" || command === "palette-rgba") {
    const rom = readFileSync(input), address = number(option(args, "--address")), size = number(option(args, "--size"));
    const start = address - ROM_BASE, end = start + size;
    if (start < 0 || end > rom.length || start >= end) throw new Error(`${command === "tiles" ? "tile" : "palette"} range is outside the ROM or empty`);
    if (same(output, input)) throw new Error("refusing to overwrite the input ROM");
    if (command === "tiles") {
      const palettePath = args.includes("--palette") ? option(args, "--palette") : undefined;
      [image, report] = tile_png(rom.subarray(start, end), number(option(args, "--bpp")), number(option(args, "--columns")), palettePath ? indexed_png(readFileSync(palettePath))[3] : undefined);
    } else [image, report] = palette_rgba_image(rom.subarray(start, end), number(args.includes("--width") ? option(args, "--width") : "16"));
  } else {
    if (same(output, input)) throw new Error("refusing to overwrite the input");
    const raw = readFileSync(input);
    if (command === "tiles-file") {
      const palettePath = args.includes("--palette") ? option(args, "--palette") : undefined;
      [image, report] = tile_png(raw, number(option(args, "--bpp")), number(option(args, "--columns")), palettePath ? indexed_png(readFileSync(palettePath))[3] : undefined);
    } else if (command === "palette-file") [image, report] = palette_png(raw);
    else if (command === "bytes-file") [image, report] = byte_png(raw, number(option(args, "--width")));
    else if (command === "rgba-file") [image, report] = rgba_image(raw, number(option(args, "--width")));
    else if (command === "palette-rgba-file") [image, report] = palette_rgba_image(raw, number(args.includes("--width") ? option(args, "--width") : "16"));
    else throw new Error("an asset command is required");
  }
  mkdirSync(dirname(output), { recursive: true }); writeFileSync(output, image); console.log(sortedJson(report));
}

if (import.meta.main) main(process.argv.slice(2));
